blockchain/indexers/common.go

```diff
@@ -173,3 +172,776 @@ func interruptRequested(ctx context.Context) bool {
+	if err := finishDrop(ctx, db, indexer); err != nil {
+		return err
+	}
+	return createIndex(db, genesisHash, indexer)
+}
+
+// IndexNtfnType represents an index notification type.
+type IndexNtfnType int
+
+const (
+	// ConnectNtfn indicates the index notification signals a block
+	// connected to the main chain.
+	ConnectNtfn IndexNtfnType = iota
+
+	// DisconnectionNtfn indicates the index notification signals a block
+	// disconnected from the main chain.
+	DisconnectNtfn
+)
+
+// IndexNtfn represents an index notification detailing a block connecting
+// or disconnecting from the main chain.
+type IndexNtfn struct {
+	NtfnType          IndexNtfnType
+	Block             *dcrutil.Block
+	Parent            *dcrutil.Block
+	PrevScripts       PrevScripter
+	IsTreasuryEnabled bool
+}
+
+// IndexSubscription represents a subscription for index updates.
+type IndexSubscription struct {
+	id         string
+	idx        Indexer
+	subscriber *IndexSubscriber
+	c          chan *IndexNtfn
+	mtx        sync.Mutex
+
+	// prerequisite defines the notification processing hierarchy for this
+	// subscription. It is expected that the subscriber associated with the
+	// prerequisite provided processes notifications before they are
+	// delivered by this subscription to its subscriber. An empty string
+	// indicates the subscription has no prerequisite.
+	prerequisite string
+
+	// dependency defines the index subscription that requires the subscriber
+	// associated with this subscription to have processed incoming
+	// notifications before it does. A nil dependency indicates the subscription
+	// has no dependencies.
+	dependency *IndexSubscription
+}
+
+// newIndexSubscription initializes a new index subscription.
+func newIndexSubscription(subber *IndexSubscriber, index Indexer, prereq string) *IndexSubscription {
+	return &IndexSubscription{
+		id:           index.Name(),
+		idx:          index,
+		prerequisite: prereq,
+		subscriber:   subber,
+		c:            make(chan *IndexNtfn, bufferSize),
+	}
+}
+
+// C returns a channel that produces a stream of index notifications.
+// Successive calls to C return the same channel.
+func (s *IndexSubscription) C() chan *IndexNtfn {
+	return s.c
+}
+
+// Stop prevents any future index updates from being delivered and
+// unsubscribes the associated subscription.
+//
+// The subscriber mutex must be held when this function is called.
+func (s *IndexSubscription) Stop() error {
+	// If the subscription has a prerequisite, find it and remove the
+	// subscription as a dependency.
+	if s.prerequisite != noPrereqs {
+		prereq, ok := s.subscriber.subscriptions[s.prerequisite]
+		if !ok {
+			return fmt.Errorf("no subscription found with id %s", s.prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		prereq.dependency = nil
+		prereq.mtx.Unlock()
+
+		return nil
+	}
+
+	// If the subscription has a dependency, stop it as well.
+	if s.dependency != nil {
+		err := s.dependency.Stop()
```

Comment:
Noted, will unexport the function. Thanks. Regarding the mutex, `Stop` currently has a single call site (at `handleNotifications` when signalling the indexes to shutdown. The mutex is held before looping through the subscriptions, since the the dependent's `Stop` is called in a recursively the initial mutex held still applies.

Reply:
I see.  I think it was the fact that it requires holding the `subscriber` mutex, and I read it as the `subscription` mutex.  That looks fine then.

This is more of a stylistic preference, but I think it would make more sense for this function to live on the `IndexSubscriber`.  Since the `IndexSubscriber` owns the `subscriptions` map, I think it should be the only one modifying it.